sdk/core/core-amqp/src/retry.ts

```diff
@@ -277,6 +277,19 @@ export async function retry<T>(config: RetryConfig<T>): Promise<T> {
   if (success) {
     return result;
   } else {
-    throw lastError;
+    throw compileErrors(errors);
   }
 }
+
+function compileErrors(errors: (MessagingError | Error)[]): MessagingError | Error {
+  if (!errors.length) {
+    throw new Error("Impossible to happen");
+  }
+  let str = "";
+  for (let i = 0; i < errors.length; ++i) {
+    str += `Error ${i}: ${errors[i]}\n\n`;
+  }
+  const lastError = errors[errors.length - 1];
+  lastError.message = str;
```

Comment:
I'm fine with this mitigation, though as an alternative/addition could we consider attaching the previous `n - 1` errors in some kind of "additionalErrors" property?

Reply:
Yeah this is related to the callback alternative I brought up in the PR description. The main difference is that the callback design enables the user to act on every error exactly when it happens so you have freedom to adjust your app behavior in real-time, e.g. stop retrying earlier. Returning a list of errors could be beneficial for custom logging but every error is already being logged by the library anyway. I am leaning towards waiting to see what customer scenarios/requirements we should base our design on.

It is also my understanding that non of the other languages do anything special to collect errors between retries, so I am glad that JS is getting a more comprehensive error message 😊